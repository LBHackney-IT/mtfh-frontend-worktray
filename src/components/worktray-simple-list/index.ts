export * from "./worktray-simple-list";
