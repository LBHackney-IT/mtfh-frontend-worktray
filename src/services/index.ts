export { default as locale } from "./locale";
export { config } from "./config";
