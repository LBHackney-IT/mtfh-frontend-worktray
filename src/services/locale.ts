const locale = {
  title: "Your tasks",
  views: {
    worktray: {
      noWorktrayResults: "No tasks found",
    },
  },
  errors: {
    unableToFetchRecord: "There was a problem retrieving the record",
    unableToFetchRecordDescription:
      "Please try again. If the issue persists, please contact support.",
    unableToFindProcess: "Invalid Process",
    unableToFindProcessDescription: "We have been unable to find a process of this type.",
    unableToFindState: "Invalid State",
    unableToFindStateDescription: "We have been unable to process a state of this type.",
  },
  components: {
    filters: {
      applyFilters: "Apply filters",
      clearFilters: "Clear filters",
      selectAll: "Select All",
      removeAll: "Remove All",
    },
    pagination: {
      previous: "Previous",
      next: "Next",
      assistiveNavigation: (page: number): string => `Page ${page}`,
      searchResultsCounter: (page: number, pageSize: number, total: number): string =>
        total === 0
          ? "Showing 0 of 0 results"
          : `Showing ${page * pageSize - pageSize + 1}â€”${Math.min(
              page * pageSize,
              total,
            )} of ${total} results`,
    },
    controls: {
      timePeriodOptions: {
        "30": "Last 30 days",
        "60": "Last 60 days",
        "3": "Last 3 months",
        "6": "Last 6 months",
        "12": "Last 12 months",
      },
      limitOption: (value: number): string => `${value} items`,
    },
  },
};

export default locale;
