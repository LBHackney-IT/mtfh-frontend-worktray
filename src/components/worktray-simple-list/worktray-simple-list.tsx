import React, { useMemo } from "react";

import { ProcessesResponse, useProcesses } from "@mtfh/common/lib/api/process/v1";
import {
  Center,
  ErrorSummary,
  SimplePagination,
  SimplePaginationButton,
  Spinner,
  Table,
  Tbody,
} from "@mtfh/common/lib/components";
import { processes as processesConfig } from "@mtfh/processes";

import { locale } from "../../services";
import { TenureProcessRecord } from "../worktray-list";

export interface WorktraySimpleListProps {
  targetId: string;
}

const NoProcesses = () => {
  return <p>No processes</p>;
};

const processRecordComponents = {
  tenure: TenureProcessRecord,
};

export const WorktraySimpleList = ({
  targetId,
}: WorktraySimpleListProps): JSX.Element => {
  const { data, size, setSize, error } = useProcesses(targetId);
  const response = useMemo<ProcessesResponse | null>(() => {
    if (!data) {
      return null;
    }
    return data[size - 1];
  }, [data, size]);

  if (error) {
    return (
      <ErrorSummary
        id="worktray-result-error"
        title={locale.errors.unableToFetchRecord}
        description={locale.errors.unableToFetchRecordDescription}
      />
    );
  }

  if (!response) {
    return (
      <Center>
        <Spinner />
      </Center>
    );
  }

  const {
    results: processes,
    paginationDetails: { nextToken },
  } = response;

  if (!processes.length) {
    return <NoProcesses />;
  }

  return (
    <div className="worktray-simple-list">
      <Table>
        <Tbody>
          {processes.map((process) => {
            const processConfig = processesConfig[process.processName];
            const Result = processRecordComponents[process.targetType];
            return (
              <Result
                key={process.id}
                process={process}
                processConfig={processConfig}
                simple
              />
            );
          })}
        </Tbody>
      </Table>
      {(size > 1 || nextToken) && (
        <SimplePagination>
          {size !== 1 && (
            <SimplePaginationButton variant="previous" onClick={() => setSize(size - 1)}>
              Previous
            </SimplePaginationButton>
          )}
          {nextToken && (
            <SimplePaginationButton variant="next" onClick={() => setSize(size + 1)}>
              Next
            </SimplePaginationButton>
          )}
        </SimplePagination>
      )}
    </div>
  );
};
