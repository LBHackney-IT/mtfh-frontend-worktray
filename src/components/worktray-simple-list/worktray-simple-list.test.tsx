import React from "react";

import {
  getTenureV1,
  mockActiveTenureV1,
  mockProcessV1,
  render,
  server,
} from "@hackney/mtfh-test-utils";
import { screen, waitFor, waitForElementToBeRemoved } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { rest } from "msw";

import { locale } from "../../services";
import { WorktraySimpleList } from "./worktray-simple-list";

const get = (path: string, data: unknown, code = 200): void => {
  server.use(
    rest.get(path, (req, res, ctx) => {
      return res(ctx.status(code), ctx.json(data));
    }),
  );
};

const mockProcessesResponse = {
  results: [
    {
      ...mockProcessV1,
      targetId: "103f3080-16d6-40fb-9273-f741ec73f7e4",
      targetType: "tenure",
      relatedEntities: [
        {
          id: "103f3080-16d6-40fb-9273-f741ec73f7e40",
          targetType: "person",
          subType: "householdMember",
          description: "Mr Ersan Kuneri",
        },
      ],
      processName: "soletojoint",
      currentState: {
        state: "DocumentsRequestedAppointment",
        createdAt: "2022-06-24T07:49:07.7892599Z",
        updatedAt: "2022-06-24T07:49:07.78926Z",
      },
    },
    {
      ...mockProcessV1,
      id: "39273cf3-6a78-45e1-32d4-b9637e310527",
      targetId: "39273cf3-6a78-45e1-32d4-b9637e310527",
      targetType: "tenure",
      relatedEntities: [
        {
          id: "39273cf3-6a78-45e1-32d4-b9637e310527",
          targetType: "person",
          subType: "householdMember",
          description: "Miss Perihan Abla",
        },
      ],
      processName: "soletojoint",
      currentState: {
        state: "AutomatedChecksPassed",
        createdAt: "2022-05-10T07:49:07.7892599Z",
        updatedAt: "2022-07-24T07:49:07.78926Z",
      },
    },
  ],
  paginationDetails: {
    nextToken: "nextToken",
  },
};

describe("WorktraySimpleList", () => {
  beforeEach(() => {
    jest.resetModules();
  });

  test("WorktraySimpleList renders correctly", async () => {
    get("/api/v1/process", mockProcessesResponse);
    server.use(getTenureV1(mockActiveTenureV1, 200));

    const { container } = render(<WorktraySimpleList targetId="123" />);
    await waitForElementToBeRemoved(screen.queryByText(/Loading/));
    expect(container).toMatchSnapshot();
  });

  test("it renders WorktraySimpleList with no results", async () => {
    get("/api/v1/process", { paginationDetails: { nextToken: null }, results: [] });
    render(<WorktraySimpleList targetId="123" />);
    await expect(screen.findByText(/No processes/)).resolves.toBeInTheDocument();
  });

  test("it pages the results", async () => {
    get("/api/v1/process", mockProcessesResponse);
    render(<WorktraySimpleList targetId="123" />);

    await expect(screen.findByText(/Next/)).resolves.toBeInTheDocument();
    await waitFor(() => expect(screen.queryByText(/Previous/)).toBe(null));

    await userEvent.click(screen.getByText(/Next/));
    await waitFor(() => expect(screen.queryByText(/Next/)).toBe(null));
    await expect(screen.findByText(/Previous/)).resolves.toBeInTheDocument();

    await userEvent.click(screen.getByText(/Previous/));
    await waitFor(() => expect(screen.queryByText(/Previous/)).toBe(null));
  });

  test("it does not render pagination unnecessarily", async () => {
    get("/api/v1/process", {
      ...mockProcessesResponse,
      paginationDetails: { nextToken: null },
    });
    render(<WorktraySimpleList targetId="123" />);

    await waitFor(() => expect(screen.queryByText(/Next/)).toBe(null));
  });

  test("it renders error", async () => {
    get("/api/v1/process", null, 500);
    render(<WorktraySimpleList targetId="123" />);

    await expect(
      screen.findByText(locale.errors.unableToFetchRecord),
    ).resolves.toBeInTheDocument();
  });
});
