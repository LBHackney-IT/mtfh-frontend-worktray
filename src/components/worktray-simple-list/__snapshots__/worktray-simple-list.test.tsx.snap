// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorktraySimpleList WorktraySimpleList renders correctly 1`] = `
<div>
  <div
    class="worktray-simple-list"
  >
    <table
      class="govuk-table lbh-table"
    >
      <tbody
        class="govuk-table__body"
      >
        <tr
          class="govuk-table__row govuk-table__row process-record"
        >
          <td
            class="govuk-table__cell process-record__item"
          >
            <a
              class="govuk-link lbh-link lbh-link--no-visited-state lbh-heading-h4"
              data-testid="person-link"
              href="/person/103f3080-16d6-40fb-9273-f741ec73f7e40"
              target="_blank"
            >
              Mr Ersan Kuneri
            </a>
            <a
              class="govuk-link lbh-link lbh-link--no-visited-state --address"
              data-testid="property-link"
              href="/property/f69ae994-4948-481f-a2d8-d52cfcb68321"
              target="_blank"
            >
              72251 Champlin Trafficway, 99647
            </a>
          </td>
          <td
            class="govuk-table__cell process-record__item --processName"
          >
            <a
              class="govuk-link lbh-link lbh-link--no-visited-state lbh-heading-h4"
              href="/processes/soletojoint/6fbe024f-2316-4265-a6e8-d65a837e308a"
              target="_blank"
            >
              Sole to Joint
            </a>
          </td>
          <td
            class="govuk-table__cell process-record__item"
          >
            <p
              class="lbh-body-m title"
            >
              Initiated
            </p>
            <p
              class="lbh-body-m createdAt"
            >
              20/08/2022
            </p>
          </td>
          <td
            class="govuk-table__cell process-record__item --status"
          >
            <span
              class="govuk-tag lbh-tag box aqua"
            >
              Awaiting supporting documents appointment
            </span>
          </td>
        </tr>
        <tr
          class="govuk-table__row govuk-table__row process-record"
        >
          <td
            class="govuk-table__cell process-record__item"
          >
            <a
              class="govuk-link lbh-link lbh-link--no-visited-state lbh-heading-h4"
              data-testid="person-link"
              href="/person/39273cf3-6a78-45e1-32d4-b9637e310527"
              target="_blank"
            >
              Miss Perihan Abla
            </a>
            <a
              class="govuk-link lbh-link lbh-link--no-visited-state --address"
              data-testid="property-link"
              href="/property/f69ae994-4948-481f-a2d8-d52cfcb68321"
              target="_blank"
            >
              72251 Champlin Trafficway, 99647
            </a>
          </td>
          <td
            class="govuk-table__cell process-record__item --processName"
          >
            <a
              class="govuk-link lbh-link lbh-link--no-visited-state lbh-heading-h4"
              href="/processes/soletojoint/39273cf3-6a78-45e1-32d4-b9637e310527"
              target="_blank"
            >
              Sole to Joint
            </a>
          </td>
          <td
            class="govuk-table__cell process-record__item"
          >
            <p
              class="lbh-body-m title"
            >
              Initiated
            </p>
            <p
              class="lbh-body-m createdAt"
            >
              20/08/2022
            </p>
          </td>
          <td
            class="govuk-table__cell process-record__item --status"
          >
            <span
              class="govuk-tag lbh-tag box aqua"
            >
              Awaiting further eligibility checks
            </span>
          </td>
        </tr>
      </tbody>
    </table>
    <nav
      class="lbh-simple-pagination"
    >
      <a
        class="lbh-simple-pagination__link lbh-simple-pagination__link--next"
        rel="next"
      >
        Next
        <svg
          fill="none"
          height="19"
          viewBox="0 0 11 19"
          width="11"
        >
          <path
            d="M1 18L9 9.5L1 1"
            stroke-width="2"
          />
        </svg>
      </a>
    </nav>
  </div>
</div>
`;
